mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -43,6 +55,379 @@ struct CompactBlockRow {
+                            ciphertext: out.enc_ciphertext[0..52].to_vec(),
+                        };
+                        outputs.push(compact_out);
+                    }
+                    // Shadowing mut variables as immutable. No longer need to update them.
+                    drop_mutability!(spends);
+                    drop_mutability!(outputs);
+                    let mut hash_tx_vec = hash_tx.0.to_vec();
+                    hash_tx_vec.reverse();
+
+                    let compact_tx = TonicCompactTx {
+                        index: tx_id as u64,
+                        hash: hash_tx_vec,
+                        fee: 0,
+                        spends,
+                        outputs,
+                    };
+                    compact_txs.push(compact_tx);
+                }
+            }
+            let mut hash = block.hash.0.to_vec();
+            hash.reverse();
+            // Set 0 in vector in the case of genesis block.
+            let mut prev_hash = block.previousblockhash.unwrap_or_default().0.to_vec();
+            prev_hash.reverse();
+            // Shadowing mut variables as immutable.
+            drop_mutability!(hash);
+            drop_mutability!(prev_hash);
+            drop_mutability!(compact_txs);
+
+            let compact_block = TonicCompactBlock {
+                proto_version: 0,
+                height,
+                hash,
+                prev_hash,
+                time: block.time,
+                // (hash, prevHash, and time) OR (full header)
+                header: Vec::new(),
+                vtx: compact_txs,
+            };
+            on_block(compact_block)?;
+        }
+        Ok(())
+    }
+
+    async fn check_watch_for_tx(&mut self, tx_id: TxId, attempts: &mut i32, watch_for_tx: &mut Option<TxId>) {
+        loop {
+            match self.get_raw_transaction_bytes(&H256Json::from(tx_id.0)).compat().await {
+                Ok(_) => break,
+                Err(e) => {
+                    error!("Error on getting tx {}", tx_id);
+                    if e.to_string().contains(NO_TX_ERROR_CODE) {
+                        if *attempts >= 3 {
+                            *watch_for_tx = None;
+                            return;
+                        }
+                        *attempts += 1;
+                    }
+                    Timer::sleep(30.).await;
+                },
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
+fn increase_delay(delay: &AtomicU64) {
+    if delay.load(AtomicOrdering::Relaxed) < 60 {
+        delay.fetch_add(5, AtomicOrdering::Relaxed);
+    }
+}
+
+macro_rules! try_loop {
+    ($e:expr, $addr: ident, $delay: ident) => {
+        match $e {
+            Ok(res) => res,
+            Err(e) => {
+                error!("{:?} error {:?}", $addr, e);
+                increase_delay(&$delay);
+                continue;
+            },
+        }
+    };
+}
+
+#[allow(dead_code)]
+struct LightwalletdConnection<C> {
```

Comment:
It seems you understood `They should work in a round-robin way, similar to how it's done in Electrum client.` too literally 🙂 I have meant that it should *work* similarly as multiple electrum servers, not that it should have the same implementation. There should be no connection loops for lightwalletd, nor it needs `JsonRpcPendingRequestsShared` as it's not JSON-RPC. As a more relevant example for this, please take a look at
https://github.com/KomodoPlatform/atomicDEX-API/blob/ebb0e81e14e4bd1e90a8068552c5e072bd5f1791/mm2src/coins/eth/web3_transport.rs#L243

It iterates over the initialized clients/nodes, tries to perform the desired request one by one, and returns the first successful response immediately. Please note that you should not *copy* the code, you should implement it yourself because lightwalletd is very different from other clients we support.

Reply:
Thanks for the note, I will change my strategy of implementation.

Reply:
@laruh Could you please remove this unused code from this PR? 🙂

Reply:
done